File: src/axon/ir_compiler.py
"""AXON Intermediate Representation (IR) compiler.

Transforms `.ax` source files into portable `.axonir` JSON by walking
the parser AST and emitting a faithful IR representation.

Also provides IR-to-AST conversion so the Python runtime can consume IR
without a massive refactor.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from axon.ir_schema import (
    AgentDef,
    Annotation,
    AxonIR,
    FlowDef,
    FlowEdge,
    ImportDef,
    MemoryDecl,
    MethodDef,
    Param,
    PromptDef,
    ProviderRef,
    RagDef,
    StageDef,
    ToolDef,
    TypeAliasDef,
)


class IRCompileError(Exception):
    """Compilation from AXON source to IR failed."""
    pass


def compile_to_ir(source: Path) -> AxonIR:
    """Compile an AXON source file into IR.

    Args:
        source: Path to the `.ax` source file.

    Returns:
        Populated ``AxonIR`` instance.

    Raises:
        IRCompileError: On parse or validation failure.
    """
    text = source.read_text(encoding="utf-8")

    from axon.parser import parse
    from axon.validator import validate

    try:
        declarations = parse(text)
    except SyntaxError as exc:
        raise IRCompileError(f"Parse error in {source}: {exc}") from exc

    diagnostics = validate(declarations)
    errors = [d for d in diagnostics if d.severity == "error"]
    if errors:
        msgs = "\n".join(str(e) for e in errors)
        raise IRCompileError(f"Validation failed:\n{msgs}")

    ir = AxonIR(version="0.2.0")
    _extract_imports(declarations, ir)
    _extract_type_aliases(declarations, ir)
    _extract_rags(declarations, ir)
    _extract_prompts(declarations, ir)
    _extract_tools(declarations, ir)
    _extract_agents(declarations, ir)
    _extract_flows(declarations, ir)
    return ir


# ---------------------------------------------------------------------------
# Extraction helpers — one per AST node type
# ---------------------------------------------------------------------------


def _annotations_from_ast(annotations: list[Any]) -> list[Annotation]:
    return [
        Annotation(name=a.name, args=dict(a.args))
        for a in annotations
    ]


def _param_from_ast(p: Any) -> Param:
    return Param(
        name=p.name,
        type_str=p.type_str,
        default=p.default,
    )


def _method_from_ast(m: Any) -> MethodDef:
    return MethodDef(
        name=m.name,
        params=[_param_from_ast(p) for p in m.params],
        return_type=m.return_type,
        body=m.body,
        annotations=_annotations_from_ast(m.annotations),
    )


def _extract_imports(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import ImportDecl
    for decl in declarations:
        if isinstance(decl, ImportDecl):
            ir.imports.append(ImportDef(
                names=list(decl.names),
                source=decl.source,
            ))


def _extract_type_aliases(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import TypeAliasDecl
    for decl in declarations:
        if isinstance(decl, TypeAliasDecl):
            ir.type_aliases.append(TypeAliasDef(
                name=decl.name,
                type_params=list(decl.type_params),
                value=decl.value,
                fields=[_param_from_ast(p) for p in decl.fields],
            ))


def _extract_rags(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import RagDecl
    for decl in declarations:
        if isinstance(decl, RagDecl):
            ir.rags.append(RagDef(
                name=decl.name,
                source=decl.source,
                chunker=decl.chunker,
                embedder=decl.embedder,
                store=decl.store,
                methods=[_method_from_ast(m) for m in decl.methods],
                annotations=_annotations_from_ast(decl.annotations),
            ))


def _extract_prompts(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import PromptDecl
    for decl in declarations:
        if isinstance(decl, PromptDecl):
            ir.prompts.append(PromptDef(
                name=decl.name,
                params=[_param_from_ast(p) for p in decl.params],
                return_type=decl.return_type,
                template=decl.template,
                annotations=_annotations_from_ast(decl.annotations),
            ))


def _extract_tools(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import ToolDecl
    for decl in declarations:
        if isinstance(decl, ToolDecl):
            ir.tools.append(ToolDef(
                name=decl.name,
                params=[_param_from_ast(p) for p in decl.params],
                return_type=decl.return_type,
                docstrings=list(decl.docstrings),
                body=decl.body,
                annotations=_annotations_from_ast(decl.annotations),
            ))


def _extract_agents(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import AgentDecl
    for decl in declarations:
        if isinstance(decl, AgentDecl):
            memory = None
            if decl.memory is not None:
                memory = MemoryDecl(
                    kind=decl.memory.kind,
                    options=dict(decl.memory.options),
                )

            ir.agents.append(AgentDef(
                name=decl.name,
                model=(decl.model or "").lstrip("@"),
                tools=list(decl.tools),
                memory=memory,
                methods=[_method_from_ast(m) for m in decl.methods],
                annotations=_annotations_from_ast(decl.annotations),
                workers=decl.workers,
            ))


def _extract_flows(declarations: list[Any], ir: AxonIR) -> None:
    from axon.ast_nodes import FlowDecl
    for decl in declarations:
        if isinstance(decl, FlowDecl):
            stages = [
                StageDef(
                    name=s.name,
                    params=[_param_from_ast(p) for p in s.params],
                    return_type=s.return_type,
                )
                for s in decl.stages
            ]

            edges: list[FlowEdge] = []
            # Parse arrow syntax from flow body, e.g. "Research -> Write"
            body = decl.body or ""
            for line in body.splitlines():
                line = line.strip()
                if "->" in line and not line.startswith("//"):
                    parts = line.split("->")
                    if len(parts) == 2:
                        from_stage = parts[0].strip()
                        to_stage = parts[1].strip()
                        if from_stage and to_stage:
                            edges.append(FlowEdge(
                                from_stage=from_stage,
                                to_stage=to_stage,
                            ))

            ir.flows.append(FlowDef(
                name=decl.name,
                params=[_param_from_ast(p) for p in decl.params],
                return_type=decl.return_type,
                stages=stages,
                edges=edges,
                annotations=_annotations_from_ast(decl.annotations),
            ))


# ---------------------------------------------------------------------------
# IR → AST conversion (enables runtime to consume .axonir files)
# ---------------------------------------------------------------------------


def _annotation_to_ast(a: Annotation) -> Any:
    from axon.ast_nodes import Annotation as ASTAnnotation
    return ASTAnnotation(name=a.name, args=dict(a.args))


def _param_to_ast(p: Param) -> Any:
    from axon.ast_nodes import Param as ASTParam
    return ASTParam(name=p.name, type_str=p.type_str, default=p.default)


def _method_to_ast(m: MethodDef) -> Any:
    from axon.ast_nodes import MethodDecl
    parsed_body = None
    if m.body.strip():
        try:
            from axon.expression_parser import parse_expression
            parsed_body = parse_expression(m.body)
        except Exception:
            pass  # fall back to None; runtime will use _infer_body_expr
    return MethodDecl(
        name=m.name,
        params=[_param_to_ast(p) for p in m.params],
        return_type=m.return_type,
        annotations=[_annotation_to_ast(a) for a in m.annotations],
        body=m.body,
        parsed_body=parsed_body,
    )


def ir_to_ast(ir: AxonIR) -> list[Any]:
    """Convert an IR document back into a list of AST declarations.

    This lets the Python runtime consume `.axonir` files without a massive
    refactor.  The resulting AST objects are functionally identical to what
    ``parse()`` would produce (minus ``line`` numbers and ``parsed_body``).
    """
    declarations: list[Any] = []

    from axon.ast_nodes import (
        AgentDecl,
        FlowDecl,
        ImportDecl,
        MemoryDecl,
        PromptDecl,
        RagDecl,
        StageDecl,
        ToolDecl,
        TypeAliasDecl,
    )

    for imp in ir.imports:
        declarations.append(ImportDecl(names=list(imp.names), source=imp.source))

    for ta in ir.type_aliases:
        declarations.append(TypeAliasDecl(
            name=ta.name,
            type_params=list(ta.type_params),
            value=ta.value,
            fields=[_param_to_ast(p) for p in ta.fields],
            line=0,
        ))

    for rag in ir.rags:
        declarations.append(RagDecl(
            name=rag.name,
            source=rag.source,
            chunker=rag.chunker,
            embedder=rag.embedder,
            store=rag.store,
            annotations=[_annotation_to_ast(a) for a in rag.annotations],
            methods=[_method_to_ast(m) for m in rag.methods],
            line=0,
        ))

    for prompt in ir.prompts:
        declarations.append(PromptDecl(
            name=prompt.name,
            params=[_param_to_ast(p) for p in prompt.params],
            return_type=prompt.return_type,
            template=prompt.template,
            annotations=[_annotation_to_ast(a) for a in prompt.annotations],
            line=0,
        ))

    for tool in ir.tools:
        declarations.append(ToolDecl(
            name=tool.name,
            params=[_param_to_ast(p) for p in tool.params],
            return_type=tool.return_type,
            docstrings=list(tool.docstrings),
            body=tool.body,
            annotations=[_annotation_to_ast(a) for a in tool.annotations],
            line=0,
            parsed_body=None,
        ))

    for agent in ir.agents:
        memory = None
        if agent.memory is not None:
            memory = MemoryDecl(kind=agent.memory.kind, options=dict(agent.memory.options))

        declarations.append(AgentDecl(
            name=agent.name,
            model=agent.model,
            tools=list(agent.tools),
            memory=memory,
            annotations=[_annotation_to_ast(a) for a in agent.annotations],
            methods=[_method_to_ast(m) for m in agent.methods],
            workers=agent.workers,
            line=0,
        ))

    for flow in ir.flows:
        # Reconstruct arrow body from edges
        arrow_lines = [f"{e.from_stage} -> {e.to_stage}" for e in flow.edges]
        body = "\n".join(arrow_lines)

        declarations.append(FlowDecl(
            name=flow.name,
            params=[_param_to_ast(p) for p in flow.params],
            return_type=flow.return_type,
            annotations=[_annotation_to_ast(a) for a in flow.annotations],
            stages=[StageDecl(
                name=s.name,
                params=[_param_to_ast(p) for p in s.params],
                return_type=s.return_type,
                line=0,
            ) for s in flow.stages],
            body=body,
            parsed_body=None,
            line=0,
        ))

    return declarations


def load_ir(source: Path) -> AxonIR:
    """Load an ``.axonir`` JSON file into an ``AxonIR`` instance."""
    import json
    text = source.read_text(encoding="utf-8")
    data = json.loads(text)
    return AxonIR.from_dict(data)

File: axon-parser/src/ir_compiler.rs
//! IR compiler — compiles AXON source to IR JSON and provides IR-to-AST conversion.
//!
//! In the Rust crate, `parse_source` already produces `AxonIR` directly,
//! so the "compile" step is simply parse + validate. This module provides
//! the `compile_source` function and IR-to-AST-JSON conversion for
//! Python interop via PyO3.

use crate::validator::{self, Diagnostic};
use crate::AxonIR;
use serde_json::{json, Value};

/// Compile AXON source text into IR, running validation.
/// Returns the IR on success, or diagnostics on validation failure.
pub fn compile_source(source: &str) -> Result<AxonIR, Vec<Diagnostic>> {
    let ir = match crate::parse_source(source) {
        Ok(ir) => ir,
        Err(e) => {
            return Err(vec![Diagnostic::error(
                format!("Parse error: {}", e), 0, "parse-error",
            )]);
        }
    };
    let diagnostics = validator::validate(&ir);
    if validator::has_errors(&diagnostics) {
        return Err(diagnostics);
    }
    Ok(ir)
}

/// Compile AXON source and return IR as JSON string.
pub fn compile_to_json(source: &str) -> Result<String, String> {
    match compile_source(source) {
        Ok(ir) => serde_json::to_string_pretty(&ir).map_err(|e| format!("Serialization error: {}", e)),
        Err(diags) => {
            let msgs: Vec<String> = diags.iter().map(|d| d.format()).collect();
            Err(msgs.join("\n"))
        }
    }
}

/// Convert IR JSON back into a JSON representation of AST declarations.
/// This mirrors the Python `ir_to_ast` function for round-trip compatibility.
pub fn ir_to_ast_json(ir_json: &str) -> Result<String, String> {
    let ir: AxonIR = serde_json::from_str(ir_json).map_err(|e| format!("IR parse error: {}", e))?;
    let declarations = ir_to_declarations_json(&ir);
    serde_json::to_string_pretty(&declarations).map_err(|e| format!("Serialization error: {}", e))
}

/// Convert IR to a JSON array of declaration dicts matching Python AST structure.
fn ir_to_declarations_json(ir: &AxonIR) -> Value {
    let mut decls: Vec<Value> = Vec::new();

    for imp in &ir.imports {
        decls.push(json!({
            "kind": "import",
            "names": imp.names,
            "source": imp.source,
        }));
    }

    for ta in &ir.type_aliases {
        decls.push(json!({
            "kind": "type_alias",
            "name": ta.name,
            "type_params": ta.type_params,
            "value": ta.value,
            "fields": ta.fields.iter().map(|p| json!({
                "name": p.name,
                "type_str": p.type_str,
                "default": p.default,
            })).collect::<Vec<_>>(),
            "line": 0,
        }));
    }

    for rag in &ir.rags {
        decls.push(json!({
            "kind": "rag",
            "name": rag.name,
            "source": rag.source,
            "chunker": rag.chunker,
            "embedder": rag.embedder,
            "store": rag.store,
            "annotations": annotations_to_json(&rag.annotations),
            "methods": methods_to_json(&rag.methods),
            "line": 0,
        }));
    }

    for prompt in &ir.prompts {
        decls.push(json!({
            "kind": "prompt",
            "name": prompt.name,
            "params": params_to_json(&prompt.params),
            "return_type": prompt.return_type,
            "template": prompt.template,
            "annotations": annotations_to_json(&prompt.annotations),
            "line": 0,
        }));
    }

    for tool in &ir.tools {
        decls.push(json!({
            "kind": "tool",
            "name": tool.name,
            "params": params_to_json(&tool.params),
            "return_type": tool.return_type,
            "docstrings": tool.docstrings,
            "body": tool.body,
            "annotations": annotations_to_json(&tool.annotations),
            "line": 0,
            "parsed_body": null,
        }));
    }

    for agent in &ir.agents {
        let memory = if let Some(mem) = &agent.memory {
            json!({"kind": mem.kind, "options": mem.options})
        } else {
            Value::Null
        };
        decls.push(json!({
            "kind": "agent",
            "name": agent.name,
            "model": agent.model,
            "tools": agent.tools,
            "memory": memory,
            "annotations": annotations_to_json(&agent.annotations),
            "methods": methods_to_json(&agent.methods),
            "workers": agent.workers,
            "version": agent.version,
            "line": 0,
        }));
    }

    for flow in &ir.flows {
        let arrow_lines: Vec<String> = flow.edges.iter()
            .map(|e| format!("{} -> {}", e.from_stage, e.to_stage))
            .collect();
        let body = arrow_lines.join("\n");
        decls.push(json!({
            "kind": "flow",
            "name": flow.name,
            "params": params_to_json(&flow.params),
            "return_type": flow.return_type,
            "annotations": annotations_to_json(&flow.annotations),
            "stages": flow.stages.iter().map(|s| json!({
                "name": s.name,
                "params": params_to_json(&s.params),
                "return_type": s.return_type,
                "line": 0,
            })).collect::<Vec<_>>(),
            "body": body,
            "parsed_body": null,
            "line": 0,
        }));
    }

    Value::Array(decls)
}

fn annotations_to_json(anns: &[crate::Annotation]) -> Vec<serde_json::Value> {
    anns.iter().map(|a| json!({
        "name": a.name,
        "args": a.args,
    })).collect()
}

fn params_to_json(params: &[crate::Param]) -> Vec<serde_json::Value> {
    params.iter().map(|p| json!({
        "name": p.name,
        "type_str": p.type_str,
        "default": p.default,
    })).collect()
}

fn methods_to_json(methods: &[crate::MethodDef]) -> Vec<serde_json::Value> {
    methods.iter().map(|m| json!({
        "name": m.name,
        "params": params_to_json(&m.params),
        "return_type": m.return_type,
        "body": m.body,
        "annotations": annotations_to_json(&m.annotations),
    })).collect()
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compile_valid_source() {
        let src = r#"
tool Add(a: Int, b: Int) -> Int {
    /// Adds two numbers
    a + b
}
"#;
        let result = compile_source(src);
        assert!(result.is_ok(), "Expected Ok, got: {:?}", result.err());
        let ir = result.unwrap();
        assert_eq!(ir.tools.len(), 1);
        assert_eq!(ir.tools[0].name, "Add");
    }

    #[test]
    fn test_compile_invalid_source() {
        let src = r#"
tool Foo(a: Int) -> Int {
    a
}
"#;
        let result = compile_source(src);
        assert!(result.is_err(), "Expected validation error");
        let diags = result.unwrap_err();
        assert!(diags.iter().any(|d| d.code == "missing-tool-docstring"));
    }

    #[test]
    fn test_compile_to_json() {
        let src = r#"
tool Add(a: Int, b: Int) -> Int {
    /// Adds two numbers
    a + b
}
"#;
        let json = compile_to_json(src).unwrap();
        assert!(json.contains("\"name\": \"Add\""));
        assert!(json.contains("\"version\""));
    }

    #[test]
    fn test_ir_to_ast_json() {
        let src = r#"
tool Add(a: Int, b: Int) -> Int {
    /// Adds two numbers
    a + b
}
"#;
        let ir_json = compile_to_json(src).unwrap();
        let ast_json = ir_to_ast_json(&ir_json).unwrap();
        assert!(ast_json.contains("\"kind\": \"tool\""));
        assert!(ast_json.contains("\"name\": \"Add\""));
    }

    #[test]
    fn test_ir_to_ast_with_agent() {
        let src = r#"
tool Greet(name: Str) -> Str {
    /// Says hello
    name
}
agent Bot {
    model: openai/gpt-4
    tools: [Greet]
    fn run(query: Str) -> Str {
        query
    }
}
"#;
        let ir_json = compile_to_json(src).unwrap();
        let ast_json = ir_to_ast_json(&ir_json).unwrap();
        assert!(ast_json.contains("\"kind\": \"agent\""));
        assert!(ast_json.contains("\"kind\": \"tool\""));
    }
}
